otter/convergence/gathering.py

```diff
@@ -111,8 +115,50 @@ def group_id(s):
                             filter(server_predicate),
                             filter(has_group_id))
 
-    eff = get_all_server_details(changes_since)
-    return eff.on(servers_apply)
+    return get_all_server_details(changes_since).on(servers_apply)
+
+
+def merge_servers(first, second):
+    """
+    Return servers got by merging the two sets of servers where second
+    takes precedence over first
+
+    :param list first: List of server dicts
+    :param list second: List of server dicts
+    :return: List of merged server dicts
+    """
+    def sdict(servers):
+        return {s['id']: s for s in servers}
+
+    return merge(sdict(first), sdict(second)).itervalues()
+
+
+@do
+def get_scaling_group_servers(tenant_id, group_id, now,
+                              cache_class=CassScalingGroupServersCache,
+                              all_as_servers=get_all_scaling_group_servers):
+    """
+    Get a group's servers taken from cache if it exists. Updates cache
+    if it is empty from newly fetched servers
+    # NOTE: This function takes tenant_id even though the whole effect is
+    # scoped on the tenant because cache calls require tenant_id. Should
+    # they also not take tenant_id and work on the scope?
+
+    :return: Servers as iterator of dicts
+    :rtype: Effect
+    """
+    cache = cache_class(tenant_id, group_id)
+    cached_servers, last_update = yield cache.get_servers()
+    if last_update is None or now - last_update >= timedelta(days=30):
+        last_update = now - timedelta(days=30)
+        changes = (yield all_as_servers(last_update))[group_id]
```

Comment:
This probably  needs to be `.get(group_id, {})`, for when there are no servers yet.